lib/identity_cache/cache_fetcher.rb

```diff
@@ -24,27 +75,179 @@ def fetch_multi(keys, &block)
+          if lock && !using_fallback_key
+            # cache invalidated during lock wait, try fallback key
+            using_fallback_key = true
+            key = lock_fill_fallback_key(key, lock)
+            expiration_options = fallback_key_expiration_options(fill_lock_duration)
+            # loop around to retry with fallback key
+          else
+            # cache invalidation prevented lock from being taken
+            return yield
+          end
+        when nil # Errors talking to memcached
+          return yield
+        else # hit
+          return result
+        end
+      end
+      raise "unexpected number of loop iterations"
+    end
+
+    def cas_or_add(key, expiration_options = EMPTY_HASH)
+      yielded = false
+      swapped = @cache_backend.cas(key, expiration_options) do |value|
+        yielded = true
+        yield value
       end
       unless yielded
-        result = yield
-        add(key, result)
+        data = yield nil
+        swapped = add(key, data, expiration_options)
       end
-      result
+      swapped
     end
 
-    private
+    def fetch_or_take_lock(key, old_lock:, **expiration_options)
+      new_lock = nil
+      swapped = cas_or_add(key, expiration_options) do |value|
+        if value.nil? || value == IdentityCache::DELETED
+          if old_lock # cache invalidated
+            return value
+          else
+            new_lock = FillLock.build(client_id: client_id, data_version: SecureRandom.uuid)
+          end
+        elsif FillLock.cache_value?(value)
+          if old_lock && old_lock.equals_cache_value?(value)
+            # preserve data version since there hasn't been any cache invalidations
+            new_lock = FillLock.build(client_id: client_id, data_version: old_lock.data_version)
+          else
+            return FillLock.from_cache(value)
+          end
+        else # hit
+          return value
+        end
+        new_lock.cache_value # take lock
+      end
+
+      return new_lock if swapped
+
+      value = @cache_backend.read(key)
+      if FillLock.cache_value?(value)
+        FillLock.from_cache(value)
+      else
+        value
+      end
+    end
+
+    def fill_with_lock(key, data, my_lock, expiration_options)
+      swapped = cas_or_add(key, expiration_options) do |value|
+        return false if value.nil? || value == IdentityCache::DELETED
+        return true unless FillLock.cache_value?(value) # already filled
+        current_lock = FillLock.from_cache(value)
+        if current_lock.data_version != my_lock.data_version
+          return false # invalidated then relocked
+        end
+        data
+      end
+
+      swapped
+    end
+
+    def lock_fill_fallback_key(key, lock)
+      "lock_fill:#{lock.data_version}:#{key}"
+    end
+
+    def fallback_key_expiration_options(fill_lock_duration)
+      # Use lower TTL for fallback key lock since it won't be used for very long.
+      expires_in = fill_lock_duration * 2
```

Comment:
The comment on this line is unclear. Lower TTL than what? The `fill_lock_duration * 2` makes it seem like this is a *higher* TTL, not a lower one.

Reply:
I'll change `Use lower ` to `Override the default`, since this just overrides the default TTL.  `fill_lock_duration` isn't normally used for a cache key TTL, it is normally used for the sleep duration.